### ISSUE TITLE
can't fetch pool data

### ISSUE THREAD

**Issue by xmonader** [threefoldtech/tfexplorer#146]
e.g 

https://explorer.testnet.grid.tf/api/v1/reservations/pools/1

https://explorer.testnet.grid.tf/api/v1/reservations/pools/124

**Comment by m-motawea** [comment]
it is happening again now. we can fetch workloads normally but pools endpoints are not responding to requests
https://explorer.testnet.grid.tf/api/v1/reservations/pools/242

**Comment by xmonader** [comment]
incident happened again today https://explorer.testnet.grid.tf/api/v1/reservations/pools/32

**Comment by LeeSmet** [comment]
goroutine stack trace:

```
goroutine profile: total 83
28 @ 0x887765 0x85309c 0x852e55 0x12c66bf 0x12d6dc2 0x125cb87 0xb51d44 0x125dc0e 0xb51d44 0xc550d3 0xb51d44 0xba9e13 0xed35a8 0xed295e 0xb55443 0xb50b4d 0x8bd0a1
#	0x12c66be	github.com/threefoldtech/tfexplorer/pkg/capacity.(*NaivePlanner).PoolsForOwner+0xde	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/capacity/planner.go:255
#	0x12d6dc1	github.com/threefoldtech/tfexplorer/pkg/workloads.(*API).listPools+0x181		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/workloads/reservation.go:270
#	0x125cb86	github.com/threefoldtech/tfexplorer/mw.AsHandlerFunc.func1+0x46				/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/action.go:28
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0x125dc0d	github.com/threefoldtech/tfexplorer/mw.(*DatabaseMiddleware).Middleware.func1+0x1cd	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/db.go:31
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xc550d2	github.com/rusart/muxprom.(*MuxProm).middleware.func1+0x1f2				/home/zaibon/dev/go/pkg/mod/github.com/rusart/muxprom@v0.0.0-20200323164249-36ea051efbe6/prom.go:137
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xba9e12	github.com/gorilla/mux.(*Router).ServeHTTP+0xd2						/home/zaibon/dev/go/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210
#	0xed35a7	github.com/gorilla/handlers.loggingHandler.ServeHTTP+0x247				/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/logging.go:45
#	0xed295d	github.com/gorilla/handlers.(*cors).ServeHTTP+0x103d					/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/cors.go:54
#	0xb55442	net/http.serverHandler.ServeHTTP+0xa2							/usr/lib/go/src/net/http/server.go:2843
#	0xb50b4c	net/http.(*conn).serve+0x8ac								/usr/lib/go/src/net/http/server.go:1925

24 @ 0x887765 0x85309c 0x852e55 0x12c649c 0x12d69b2 0x125cb87 0xb51d44 0x125dc0e 0xb51d44 0xc550d3 0xb51d44 0xba9e13 0xed35a8 0xed295e 0xb55443 0xb50b4d 0x8bd0a1
#	0x12c649b	github.com/threefoldtech/tfexplorer/pkg/capacity.(*NaivePlanner).PoolByID+0xfb		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/capacity/planner.go:237
#	0x12d69b1	github.com/threefoldtech/tfexplorer/pkg/workloads.(*API).getPool+0x1d1			/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/workloads/reservation.go:251
#	0x125cb86	github.com/threefoldtech/tfexplorer/mw.AsHandlerFunc.func1+0x46				/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/action.go:28
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0x125dc0d	github.com/threefoldtech/tfexplorer/mw.(*DatabaseMiddleware).Middleware.func1+0x1cd	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/db.go:31
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xc550d2	github.com/rusart/muxprom.(*MuxProm).middleware.func1+0x1f2				/home/zaibon/dev/go/pkg/mod/github.com/rusart/muxprom@v0.0.0-20200323164249-36ea051efbe6/prom.go:137
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xba9e12	github.com/gorilla/mux.(*Router).ServeHTTP+0xd2						/home/zaibon/dev/go/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210
#	0xed35a7	github.com/gorilla/handlers.loggingHandler.ServeHTTP+0x247				/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/logging.go:45
#	0xed295d	github.com/gorilla/handlers.(*cors).ServeHTTP+0x103d					/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/cors.go:54
#	0xb55442	net/http.serverHandler.ServeHTTP+0xa2							/usr/lib/go/src/net/http/server.go:2843
#	0xb50b4c	net/http.(*conn).serve+0x8ac								/usr/lib/go/src/net/http/server.go:1925

20 @ 0x887765 0x88007b 0x8b7655 0x92a865 0x92b8b1 0x92b893 0xa18aef 0xa2cd4e 0xb4a8d8 0x8bd0a1
#	0x8b7654	internal/poll.runtime_pollWait+0x54		/usr/lib/go/src/runtime/netpoll.go:220
#	0x92a864	internal/poll.(*pollDesc).wait+0x44		/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87
#	0x92b8b0	internal/poll.(*pollDesc).waitRead+0x1b0	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
#	0x92b892	internal/poll.(*FD).Read+0x192			/usr/lib/go/src/internal/poll/fd_unix.go:159
#	0xa18aee	net.(*netFD).Read+0x4e				/usr/lib/go/src/net/fd_posix.go:55
#	0xa2cd4d	net.(*conn).Read+0x8d				/usr/lib/go/src/net/net.go:182
#	0xb4a8d7	net/http.(*connReader).backgroundRead+0x57	/usr/lib/go/src/net/http/server.go:690

3 @ 0x887765 0x8978a5 0xe16db8 0x8bd0a1
#	0xe16db7	go.mongodb.org/mongo-driver/x/mongo/driver/topology.(*Server).update+0x417	/home/zaibon/dev/go/pkg/mod/go.mongodb.org/mongo-driver@v1.3.2/x/mongo/driver/topology/server.go:420

1 @ 0x8599f4 0x8b985d 0xb86065 0x8bd0a1
#	0x8b985c	os/signal.signal_recv+0x9c	/usr/lib/go/src/runtime/sigqueue.go:147
#	0xb86064	os/signal.loop+0x24		/usr/lib/go/src/os/signal/signal_unix.go:23

1 @ 0x887765 0x85309c 0x852e55 0x12be8c9 0x12bd1f8 0x12bc27a 0x8bd0a1
#	0x12be8c8	github.com/threefoldtech/tfexplorer/pkg/escrow.(*Stellar).checkCapacityReservationPaid+0x7e8	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/escrow/stellar.go:404
#	0x12bd1f7	github.com/threefoldtech/tfexplorer/pkg/escrow.(*Stellar).checkCapacityReservations+0x157	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/escrow/stellar.go:278
#	0x12bc279	github.com/threefoldtech/tfexplorer/pkg/escrow.(*Stellar).Run+0xbf9				/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/escrow/stellar.go:142

1 @ 0x887765 0x85309c 0x852e55 0x12c3e65 0x12c6dc5 0x12c5b4b 0x8bd0a1
#	0x12c3e64	github.com/threefoldtech/tfexplorer/pkg/escrow.(*Stellar).CapacityReservation+0x164	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/escrow/stellar.go:881
#	0x12c6dc4	github.com/threefoldtech/tfexplorer/pkg/capacity.(*NaivePlanner).reserve+0x3a4		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/capacity/planner.go:338
#	0x12c5b4a	github.com/threefoldtech/tfexplorer/pkg/capacity.(*NaivePlanner).Run+0x8ca		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/capacity/planner.go:155

1 @ 0x887765 0x853d65 0x8539ab 0x12c6005 0x12d5c45 0x125cb87 0xb51d44 0x125dc0e 0xb51d44 0xc550d3 0xb51d44 0xba9e13 0xed35a8 0xed295e 0xb55443 0xb50b4d 0x8bd0a1
#	0x12c6004	github.com/threefoldtech/tfexplorer/pkg/capacity.(*NaivePlanner).Reserve+0x1a4		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/capacity/planner.go:196
#	0x12d5c44	github.com/threefoldtech/tfexplorer/pkg/workloads.(*API).setupPool+0x924		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/pkg/workloads/reservation.go:232
#	0x125cb86	github.com/threefoldtech/tfexplorer/mw.AsHandlerFunc.func1+0x46				/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/action.go:28
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0x125dc0d	github.com/threefoldtech/tfexplorer/mw.(*DatabaseMiddleware).Middleware.func1+0x1cd	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/db.go:31
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xc550d2	github.com/rusart/muxprom.(*MuxProm).middleware.func1+0x1f2				/home/zaibon/dev/go/pkg/mod/github.com/rusart/muxprom@v0.0.0-20200323164249-36ea051efbe6/prom.go:137
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xba9e12	github.com/gorilla/mux.(*Router).ServeHTTP+0xd2						/home/zaibon/dev/go/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210
#	0xed35a7	github.com/gorilla/handlers.loggingHandler.ServeHTTP+0x247				/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/logging.go:45
#	0xed295d	github.com/gorilla/handlers.(*cors).ServeHTTP+0x103d					/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/cors.go:54
#	0xb55442	net/http.serverHandler.ServeHTTP+0xa2							/usr/lib/go/src/net/http/server.go:2843
#	0xb50b4c	net/http.(*conn).serve+0x8ac								/usr/lib/go/src/net/http/server.go:1925

1 @ 0x887765 0x853d65 0x8539ab 0x12ec0a7 0x887369 0x8bd0a1
#	0x12ec0a6	main.main+0x986		/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/cmds/tfexplorer/main.go:106
#	0x887368	runtime.main+0x208	/usr/lib/go/src/runtime/proc.go:204

1 @ 0x887765 0x88007b 0x8b7655 0x92a865 0x92d45c 0x92d43e 0xa1a065 0xa36a12 0xa357e5 0xb55806 0xb55537 0x8bd0a1
#	0x8b7654	internal/poll.runtime_pollWait+0x54		/usr/lib/go/src/runtime/netpoll.go:220
#	0x92a864	internal/poll.(*pollDesc).wait+0x44		/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87
#	0x92d45b	internal/poll.(*pollDesc).waitRead+0x1fb	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
#	0x92d43d	internal/poll.(*FD).Accept+0x1dd		/usr/lib/go/src/internal/poll/fd_unix.go:394
#	0xa1a064	net.(*netFD).accept+0x44			/usr/lib/go/src/net/fd_unix.go:172
#	0xa36a11	net.(*TCPListener).accept+0x31			/usr/lib/go/src/net/tcpsock_posix.go:139
#	0xa357e4	net.(*TCPListener).Accept+0x64			/usr/lib/go/src/net/tcpsock.go:261
#	0xb55805	net/http.(*Server).Serve+0x265			/usr/lib/go/src/net/http/server.go:2937
#	0xb55536	net/http.(*Server).ListenAndServe+0xb6		/usr/lib/go/src/net/http/server.go:2866

1 @ 0x8b71fd 0xec4f82 0xec4d45 0xec1912 0xecf6e5 0xed0fc5 0xb51d44 0xb53d6d 0x125dc0e 0xb51d44 0xc550d3 0xb51d44 0xba9e13 0xed35a8 0xed295e 0xb55443 0xb50b4d 0x8bd0a1
#	0x8b71fc	runtime/pprof.runtime_goroutineProfileWithLabels+0x5c					/usr/lib/go/src/runtime/mprof.go:716
#	0xec4f81	runtime/pprof.writeRuntimeProfile+0xe1							/usr/lib/go/src/runtime/pprof/pprof.go:724
#	0xec4d44	runtime/pprof.writeGoroutine+0xa4							/usr/lib/go/src/runtime/pprof/pprof.go:684
#	0xec1911	runtime/pprof.(*Profile).WriteTo+0x3f1							/usr/lib/go/src/runtime/pprof/pprof.go:331
#	0xecf6e4	net/http/pprof.handler.ServeHTTP+0x384							/usr/lib/go/src/net/http/pprof/pprof.go:256
#	0xed0fc4	net/http/pprof.Index+0x944								/usr/lib/go/src/net/http/pprof/pprof.go:367
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xb53d6c	net/http.(*ServeMux).ServeHTTP+0x1ac							/usr/lib/go/src/net/http/server.go:2417
#	0x125dc0d	github.com/threefoldtech/tfexplorer/mw.(*DatabaseMiddleware).Middleware.func1+0x1cd	/home/zaibon/dev/go/src/github.com/threefoldtech/tfexplorer/mw/db.go:31
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xc550d2	github.com/rusart/muxprom.(*MuxProm).middleware.func1+0x1f2				/home/zaibon/dev/go/pkg/mod/github.com/rusart/muxprom@v0.0.0-20200323164249-36ea051efbe6/prom.go:137
#	0xb51d43	net/http.HandlerFunc.ServeHTTP+0x43							/usr/lib/go/src/net/http/server.go:2042
#	0xba9e12	github.com/gorilla/mux.(*Router).ServeHTTP+0xd2						/home/zaibon/dev/go/pkg/mod/github.com/gorilla/mux@v1.7.4/mux.go:210
#	0xed35a7	github.com/gorilla/handlers.loggingHandler.ServeHTTP+0x247				/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/logging.go:45
#	0xed295d	github.com/gorilla/handlers.(*cors).ServeHTTP+0x103d					/home/zaibon/dev/go/pkg/mod/github.com/gorilla/handlers@v1.4.2/cors.go:54
#	0xb55442	net/http.serverHandler.ServeHTTP+0xa2							/usr/lib/go/src/net/http/server.go:2843
#	0xb50b4c	net/http.(*conn).serve+0x8ac								/usr/lib/go/src/net/http/server.go:1925

1 @ 0xb4a881 0x8bd0a1
#	0xb4a880	net/http.(*connReader).backgroundRead+0x0	/usr/lib/go/src/net/http/server.go:689
```

**Comment by LeeSmet** [comment]
So it seems the escrow worker was looking for paid reservations as it does every minute, and found one. If it does it sends the capacity reservation id on a channel, which is read by the planner. This blocks the escrow until it can send the value (since go channels aren't buffered). At the same time, someone reserved a capacity pool. This caused the planner to call the escrow, which eventually hits the worker there as well. but the worker is already blocked in the escrow, so it then blocks the planner worker as well.

**Comment by LeeSmet** [comment]
So there seem to be a couple of solutions:
1) buffer this particular channel. This increases memory consumption by a bit and its technically possible to fill a buffer as well, but in practice, setting the buffer to 100+ wil cause us other problems if it is filled in one sweep (such as stellar horizon rate limit).
2) spin up the goroutine to perform the channel write. This means we don't block the escrow  worker, so it will continue processing, eventually picking up the capacity reservation from the planner, unblocking the planner, allowing it to read the values from the paid channel. This means that we temporarily have some goroutines lying around (if there is another issue, we will stack up goroutines). It also might cause the reservatoin ids to be received out of order, not sure if this is a problem.
3) Create a cache for the paid reservations. When writing on the channel, do so in a select, with a default option to store the id in the cache slice. The main worker loop can then pull from the cache when its idling, repeating the same procedure: write in a select, with an empty default (i.e. only write if non blocking). This seems to be a slightly nicer, yet way more convoluted approach to the first option.

All in all, I think option 1 will be the best

**Comment by hossnys** [comment]
verified 

![Screenshot from 2020-08-26 15-23-06](https://user-images.githubusercontent.com/26298009/91309754-4751eb00-e7b1-11ea-8e13-e2c21c5e6e00.png)
